keyman_trust_darwin.go

```diff
@@ -20,46 +20,53 @@ func DeleteTrustedRootByName(commonName string, prompt string) error {
 	return nil
 }
 
-func (cert *Certificate) AddAsTrustedRoot(prompt string) error {
+// isInstalled checks whether this certificate is install based purely on looking for a cert
+// in the system keychain that has the same common name.  This function returns
+// true if there are one or more certs in the system keychain whose common name
+// matches this cert.
+func (cert *Certificate) isInstalled() bool {
+	cmd := exec.Command("security", "find-certificate", "-c", cert.X509().Subject.CommonName, OSX_SYSTEM_KEYCHAIN_PATH)
+	err := cmd.Run()
+
+	return err == nil
+}
+
+// AddAsTrustedRootIfNeeded adds the certificate to the user's trust store as a trusted
+// root CA.
+// elevatePrompt will be displayed when asking for admin permissions
+// installPromptTitle/Content are ignored
+// returns true if any actual changes were made
+func (cert *Certificate) AddAsTrustedRootIfNeeded(elevatePrompt, installPromptTitle, installPromptContent string) (bool, error) {
+	if cert.isInstalled() {
+		return false, nil
+	}
 	tempFileName, err := cert.WriteToTempFile()
 	defer func() {
 		if err := os.Remove(tempFileName); err != nil {
 			log.Debugf("Unable to remove file: %v", err)
 		}
 	}()
 	if err != nil {
-		return fmt.Errorf("Unable to create temp file: %s", err)
+		return false, fmt.Errorf("Unable to create temp file: %s", err)
 	}
 
 	cmd := exec.Command("security", "verify-cert", "-c", tempFileName)
 	_, err = cmd.CombinedOutput()
 	if err == nil {
 		// certificate verified successfully so it's already a trusted root, no need
 		// to install.
-		return nil
+		return false, nil
 	}
 
 	// Add it as a trusted cert
-	cmd = elevatedIfNecessary(prompt)("security", "add-trusted-cert", "-d", "-k", OSX_SYSTEM_KEYCHAIN_PATH, tempFileName)
+	cmd = elevatedIfNecessary(elevatePrompt)("security", "add-trusted-cert", "-d", "-k", OSX_SYSTEM_KEYCHAIN_PATH, tempFileName)
 	out, err := cmd.CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("Unable to run security command: %s\n%s", err, out)
+		return false, fmt.Errorf("Unable to run security command: %s\n%s", err, out)
```

Comment:
Sorry I didn't explain what I mean very well, but this needs to be true, because what the caller needs to know is whether we attempted to install the certificate, not whether we did install the certificate. Specifically, we want to record the success rate of installing the certificates, so we need to know when an attempt was made. If we attempted and the error is nil, the caller can assume that it installed successfully.

I think passing an `installAttempted` callback is probably more clear.

Reply:
@oxtoacart if I return true here, I think this should handle the use case you describe. why do you think adding `installAttempted` callback would be beneficial?

Reply:
In the case of linux, a simple true/false is a bit ambiguous. We may successfully install the certificate in 2 places but fail in the third, in which case we would return true and consider it failure, but in reality it was only a partial failure.